vllm/engine/llm_engine.py

```diff
@@ -339,6 +336,180 @@ def step(self) -> List[RequestOutput]:
+            for seq, parent, is_new in all_finished_seqs[beam_width:]:
+                if is_new:
+                    unselected_child_seqs.append((seq, parent))
+                else:
+                    seq_group.remove(seq.seq_id)
+
+            # select the top beam_width sequences from the running
+            # sequences for the next iteration to continue the beam
+            # search.
+            running_child_seqs = [(seq, parent) for seq, parent in child_seqs
+                                  if not seq.is_finished()]
+            running_child_seqs.sort(
+                key=lambda x: x[0].get_beam_search_score(length_penalty),
+                reverse=True)
+
+            # Check if we can stop the beam search.
+            if len(running_child_seqs) == 0:
+                # No running sequences, stop the beam search.
+                stop_beam_search = True
+            elif len(all_finished_seqs) < beam_width:
+                # Not enough finished sequences, continue the beam search.
+                stop_beam_search = False
+            else:
+                # Check the early stopping criteria
+                best_running_seq = running_child_seqs[0][0]
+                current_worst_seq = all_finished_seqs[beam_width - 1][0]
+                stop_beam_search = self._check_beam_search_early_stopping(
+                    seq_group.sampling_params.early_stopping,
+                    seq_group.sampling_params, best_running_seq,
+                    current_worst_seq)
+
+            if stop_beam_search:
+                unselected_child_seqs.extend(running_child_seqs)
+            else:
+                selected_child_seqs.extend(running_child_seqs[:beam_width])
+                unselected_child_seqs.extend(running_child_seqs[beam_width:])
+
+        else:
+            # Non-beam search decoding. Keep all child sequences.
+            selected_child_seqs = child_seqs
+
+        # For newly created child sequences, add them to the sequence group
+        # and fork them in block manager if they are not finished.
+        for seq, parent in selected_child_seqs:
+            if seq is not parent:
+                seq_group.add(seq)
+                if not seq.is_finished():
+                    self.scheduler.fork_seq(parent, seq)
+
+        # Free the finished and selected parent sequences' memory in block
+        # manager. Keep them in the sequence group as candidate output.
+        for seq, parent in selected_child_seqs:
+            if seq is parent and seq.is_finished():
+                self.scheduler.free_seq(seq)
+
+        # Remove the unselected parent sequences from the sequence group and
+        # free their memory in block manager.
+        for seq, parent in unselected_child_seqs:
+            if seq is parent:
+                # Remove the parent sequence if it is not selected for next
+                # iteration
+                seq_group.remove(seq.seq_id)
+                self.scheduler.free_seq(seq)
```

Comment:
Move this part into the big if above.

Reply:
Yes, I think it'd be nice if we can have a simpler version of the method for non-beam search cases, so that people who are not interested in beam search can easily follow the code,